File: 32.cpp
/**
32. Minimum Window Substring
https://www.lintcode.com/problem/minimum-window-substring/
**/
class Solution {
public:
    /**
     * @param source : A string
     * @param target: A string
     * @return: A string denote the minimum window, return "" if there is no such a string
     */
    string minWindow(string &source , string &target) {
        // write your code here
        int freqT[256] = {0};
        int unique_target_char = 0;
        for (char& c: target) {
            if (++freqT[c] == 1) ++unique_target_char;
        }
        int r = 0, n = source.size();
        int r_l = 0, r_r = n;
        int freqS[256] = {0};
        int matched_char = 0;
        for (int l = 0; l < n; ++l) {
            while (r < n && matched_char < unique_target_char) {
                if (++freqS[source[r]] == freqT[source[r]]) ++matched_char;
                ++r;
            }
            if (matched_char == unique_target_char && (r - 1 - l < r_r - r_l)) {
                r_l = l;
                r_r = r - 1;
            }
            if (--freqS[source[l]] == freqT[source[l]] - 1) --matched_char;
        }
        return r_r != n ? source.substr(r_l, r_r - r_l + 1) : "";
    }
};

File: lintcode/32.py
"""
32. Minimum Window Substring

"""
class Solution:
    """
    @param source : A string
    @param target: A string
    @return: A string denote the minimum window, return "" if there is no such a string
    """
    def minWindow(self, source , target):
        # write your code here
        freqT = [0] * 256
        target_match = 0
        for c in target:
            freqT[ord(c)] += 1
            if freqT[ord(c)] == 1:
                target_match += 1

        n = len(source)

        right = 0
        freqS = [0] * 256
        source_match = 0
        min_left, min_right = 0, n + 1

        for left in range(n):
            while right < n and source_match < target_match:
                freqS[ord(source[right])] += 1
                if freqS[ord(source[right])] == freqT[ord(source[right])]:
                    source_match += 1 #here
                right += 1
            """
            | abcdefghi | jk
              ^           ^
              l           r
            """
            if source_match == target_match and right - left < min_right - min_left:
                min_left, min_right = left, right

            freqS[ord(source[left])] -= 1
            if freqS[ord(source[left])] == freqT[ord(source[left])] - 1:
                source_match -= 1

        return source[min_left: min_right] if (min_left, min_right) != (0, n + 1) else "" #don't need + 1 because right is already + 1
